source/Octopus.Tentacle/Kubernetes/KubernetesPodMonitor.cs

```diff
@@ -85,82 +106,133 @@ async Task UpdateLoop(CancellationToken cancellationToken)
 
         internal async Task<string> InitialLoadAsync(CancellationToken cancellationToken)
         {
-            log.Verbose("Preloading pod statuses");
-
-            var newStatuses = new ConcurrentDictionary<ScriptTicket, TrackedScriptPod>();
-            var allPods = await podService.ListAllPods(cancellationToken);
-            foreach (var pod in allPods.Items)
-            {
-                var scriptTicket = pod.GetScriptTicket();
-                if (!podStatusLookup.TryGetValue(scriptTicket, out var status))
-                {
-                    status = new TrackedScriptPod(scriptTicket);
-                }
-                status.Update(pod);
-
-                log.Verbose($"Preloaded pod {pod.Name()} ({status})");
-                newStatuses[scriptTicket] = status;
-            }
-
-            // Updating a reference is an atomic operation
-            // and we only add data to this dictionary
-            // in this class so this is thread safe.
-            podStatusLookup = newStatuses;
-
-            log.Verbose($"Preloaded {allPods.Items.Count} pod statuses. ResourceVersion: {allPods.ResourceVersion()}");
-
-            //this is the resource version for the list. We use this to start the watch at this particular point
-            return allPods.ResourceVersion();
+           log.Verbose("Loading pod statuses");
+
+           Stopwatch stopwatch = Stopwatch.StartNew();
+
+           var newStatuses = new ConcurrentDictionary<ScriptTicket, TrackedScriptPod>();
+           var allPods = await podService.ListAllPods(cancellationToken);
+           foreach (var pod in allPods.Items)
+           {
+               var scriptTicket = pod.GetScriptTicket();
+               if (!newStatuses.TryGetValue(scriptTicket, out var status))
+               {
+                   status = new TrackedScriptPod(scriptTicket);
+               }
+
+               status.Update(pod);
+
+               log.Verbose($"Loaded pod {pod.Name()} ({status})");
+               newStatuses[scriptTicket] = status;
+           }
+
+           lock (theLock)
```

Comment:
I think we actually need to use a lock, since we don't want to be modifying the collections while the other methods are in the middle of reading

Reply:
@APErebus happy to talk through this when you are free